### DISCUSSION TITLE
Performance tests use liveValue

### DISCUSSION THREAD

**Discussion by mplorentz** [pointfreeco/swift-dependencies#114] [General]
### Description

I've set up a suite of UI test for performance measurements as described in the WWDC video [Elimnate animation hitches with XCTest](https://developer.apple.com/wwdc20/10077). Apple recommends setting your configuration to Release and disabling the debugger for these tests.

I want the Dependencies to use the `testValue` of my dependencies during these tests.

### Checklist

- [ ] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [ ] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-dependencies/issues) or [discussion](https://github.com/pointfreeco/swift-dependencies/discussions).

### Expected behavior

Dependencies library detects that I am running tests and initializes the `testValue` for my dependencies.

### Actual behavior

The Dependencies library initializes the `liveValue` of my dependencies in these test.

### Steps to reproduce

1. Open an Xcode project for an iOS application that uses `testValue` and `liveValue` dependencies.
1. Set up a UI test in a test target that runs against your application in release mode without the debugger attached.
2. Run the UI tests.

### Dependencies version information

0.1.4

### Destination operating system

iOS 16.4

### Xcode version information

Xcode 14.3

### Swift Compiler version information

```shell
swift-driver version: 1.75.2 Apple Swift version 5.8.1 (swiftlang-5.8.0.124.5 clang-1403.0.22.11.100)
Target: arm64-apple-macosx13.0
```

**Comment by mbrandonw** [comment]
Hi @mplorentz, this doesn't have so much to do with tests run in RELEASE vs DEBUG, but rather it's a problem with UI tests in general. A UI test runs as a completely separate process, independent of the test target, and so there is no way (as far as we know) to detect when an app is running as apart of a UI test.

So, you have to come up with your own strategies for communicating to the app target from the test target that you are running tests, and then do some extra logic in the entry point of the app target to override dependencies. For example, [this](https://github.com/pointfreeco/standups/blob/9e178eda6d8b21e92d5c5afa43dadfb355926205/Standups-StackBased/Standups/StandupsApp.swift#L12-L13) is what we do in our Standups demo app to show how to use a mock data manager in a UI test. That environment variable is [set](https://github.com/pointfreeco/standups/blob/9e178eda6d8b21e92d5c5afa43dadfb355926205/Standups-StackBased/StandupsUITests/StandupsListUITests.swift#L16-L18) from the UI test target.

Since this isn't an issue with the library I am going to convert it to a discussion. Feel free to follow up there with more questions.

**Reply by jshier** [reply]
> So, you have to come up with your own strategies for communicating to the app target from the test target that you are running tests, and then do some extra logic in the entry point of the app target to override dependencies. For example, [this](https://github.com/pointfreeco/standups/blob/9e178eda6d8b21e92d5c5afa43dadfb355926205/Standups-StackBased/Standups/StandupsApp.swift#L12-L13) is what we do in our Standups demo app to show how to use a mock data manager in a UI test. That environment variable is [set](https://github.com/pointfreeco/standups/blob/9e178eda6d8b21e92d5c5afa43dadfb355926205/Standups-StackBased/StandupsUITests/StandupsListUITests.swift#L16-L18) from the UI test target.

Perhaps this approach could be standardized within `swift-dependencies` so that developers only need to provide the environment value to activate the test dependencies? Seems like a standardized approach would be beneficial.

**Comment by mplorentz** [comment]
I played around with this some more and found the latest version of swift-dependencies [does support](https://github.com/pointfreeco/swift-dependencies/blob/6c0774bbbdc6906aa9c6bf04bb43526de3486e1f/Integration/IntegrationUITests/IntegrationUITests.swift#L11) the environment variable `SWIFT_DEPENDENCIES_CONTEXT` for this purpose! This solves my issue, so I'm going the close the discussion. It would be nice to see this in the official documentation, but now it's documented here at least for search engine crawlers.

**Reply by mbrandonw** [reply]
Honestly I even forgot we supported that option. I would love to document it, just have to figure out where it would have actually been discoverable to help in the first place.

**Reply by mplorentz** [reply]
> just have to figure out where it would have actually been discoverable to help in the first place.

Not sure if this is helps, but I found it by searching through the swift-dependencies code to see how it picked which version of the dependency to use. I was thinking about forking and adding a variable like this myself, but then saw it was already there. You can see the code for it [here](https://github.com/pointfreeco/swift-dependencies/blob/6c0774bbbdc6906aa9c6bf04bb43526de3486e1f/Integration/IntegrationUITests/IntegrationUITests.swift#L11).

**Reply by mbrandonw** [reply]
I just opened this PR https://github.com/pointfreeco/swift-dependencies/pull/115 to add some docs. Don't know if this is the best spot, but it's something for now.
